images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/polar_cloud.py

```diff
@@ -0,0 +1,274 @@
+"""
+Module to allow printing using Polar Cloud service.
+"""
+
+import asyncio
+import logging
+import os
+import subprocess
+import socketio
+from Crypto.PublicKey import RSA
+from Crypto.Cipher import PKCS1_OAEP
+from base64 import b64encode
+from typing import Dict
+
+
+# from x1plusd.dbus import
+import x1plus
+from x1plus.utils import get_MAC, get_IP, serial_number, is_emulating
+
+logger = logging.getLogger(__name__)
+
+
+class PolarPrintService:
+    def __init__(self, settings):
+        self.polar_sn = 0
+        # Todo: VERY IMPORTANT!! The public and private keys MUST be moved to
+        # non-volatile memory before release.
```

Comment:
How are these managed in the long term?  I assume they are generated on the printer, per-printer?  I could imagine either storing them in a file in `/mnt/sdcard/x1plus/printers/.../`, or maybe they should just live in the `x1plus.settings` store directly?

There is also very limited non-volatile configuration storage on eMMC (printer internal storage), but, until now, we have avoided trying to write to this.  If you think this needs to live on printer instead of on SD card, we probably should think long and hard about that, though.

Reply:
Yes, this is per-printer. I was figuring I'd put it on the SD card; it didn't occur to me that it could go in main printer memory.

I _think_ it wouldn't matter if it was per SD card, but let me talk to my colleagues and see what they think.